spikeinterface/sortingcomponents/waveforms/pca.py

```diff
@@ -0,0 +1,150 @@
+        peaks = select_peaks(peaks, **peak_selection_params) # How to select n_peaks
+
+        # Create a waveform extractor
+        ms_before = self._kwargs["ms_before"]
+        ms_after = self._kwargs["ms_after"]
+        # Creates a numpy sorting object where the spike times are the peak times and the unit ids are the peak channel
+        sorting = NumpySorting.from_peaks(peaks, sampling_frequency=recording.sampling_frequency) 
+        we = extract_waveforms(recording, sorting, ms_before=ms_before, ms_after=ms_after, folder=None, 
+                               mode="memory", max_spikes_per_unit=None, **job_kwargs)
+
+        # compute PCA by_channel_global (with sparsity)
+        sparsity = ChannelSparsity.from_radius(we, radius_um=self.local_radius_um)
+        pc = compute_principal_components(we, n_components=n_components,  mode='by_channel_global',
+                                        sparsity=sparsity, whiten=whiten)
+        self.pca_model  = pc.get_pca_model()
+
+        # model_folder should be an input
+        pca_model_path = self._kwargs["model_path"]
+
+        with open(pca_model_path, "wb") as f:
+            pickle.dump(self.pca_model, f)
+            
+        return self.pca_model
+    
+                
+class PCAProjection(PCBaseNode):
+    
+    def __init__(self, recording: BaseRecording, name="pca_projection", parents: list = None, ms_before: float = 1., ms_after: float = 1.,
+        peak_sign: str = 'neg', all_channels: bool = True, model_path: str = None, local_radius_um: float = None,
+        return_ouput=True,
+        ):
+        
+        PCBaseNode.__init__(self, recording=recording, name=name, parents=parents, ms_before=ms_before, ms_after=ms_after,
+                            peak_sign=peak_sign, all_channels=all_channels, model_path=model_path, 
+                            local_radius_um=local_radius_um, return_ouput=return_ouput)
+            
+    def compute(self, traces, peaks, waveforms):
+
+        if self.pca_model == None:
+            exception_string = (
+                f"Pca model not found, "
+                f"Train the pca model before running compute_buffer using {self.__name__}.fit(kwargs)"
+            )
+            raise AttributeError(exception_string)
+        
+        n_waveforms, n_samples , n_channels = waveforms.shape
+        n_components = self.pca_model.n_components
+        
+        # Initializing non-valid channels with nan (sparsity is used to indicate valid channels)
+        projected_waveforms = np.full(shape =(n_waveforms, n_components, n_channels), fill_value=np.nan)
+
+        for waveform_index, main_channel in enumerate(peaks['channel_ind']):
+            channel_indexes_sparsified,  = np.nonzero(self.neighbours_mask[main_channel])
+            sparsified_waveform = waveforms[waveform_index, :, channel_indexes_sparsified]
+            projected_waveforms[waveform_index, :, channel_indexes_sparsified] = self.pca_model.transform(sparsified_waveform)
```

Comment:
Did we agree to use `max_sparse_channels` and place all valid channels in the first positions?

Reply:
Yes, that's something I need to do.

Reply:
OK, I added a test for this when the waveforms are sparse.